### PR TITLE
Add MED wrapper

### PR DESCRIPTION
**Pull request by dancrepeau**
Hello,

This is the third attempt to add MED format (medformat.org) support into Neo.  After the previous pull request, my colleague Matt Stead and I re-worked the code for determining Neo streams, and addressed all the comments from the second pull request.  We also verified that multiple MED wrappers can point to the same dataset.

Please let us know if there any questions or comments!  Thanks.

### CODE REVIEW

File: neo/test/rawiotest/test_medrawio.py
Comment (JuliaSprenger) [inline neo/test/rawiotest/test_medrawio.py:28]: This test causes a Segmentation fault on github actions as well as locally for me on ubuntu 20.04. On what system did this test run successfully for you?

File: neo/test/iotest/test_medio.py
Comment (JuliaSprenger) [inline neo/test/iotest/test_medio.py:55]: The `isinstance` checks here are already covered by the neo core library and don't need to be repeated here.
```suggestion
        for anasig in seg.analogsignals:
            self.assertNotEqual(anasig.size, 0)
        for st in seg.spiketrains:
            self.assertNotEqual(st.size, 0)
```

File: neo/test/rawiotest/test_medrawio.py
Comment (JuliaSprenger) [inline neo/test/rawiotest/test_medrawio.py:46]:
```suggestion
        # Verify if found the names of the 3 channels
```

File: neo/test/rawiotest/test_medrawio.py
Comment (JuliaSprenger) [inline neo/test/rawiotest/test_medrawio.py:159]: This line is causing a Segmentation error down the line in the med package for my local runs. For the traceback of the segmentation fault on github actions, see https://github.com/NeuralEnsemble/python-neo/actions/runs/5095010482/jobs/9205183196?pr=1288

File: neo/test/rawiotest/test_medrawio.py
Comment (JuliaSprenger) [inline neo/test/rawiotest/test_medrawio.py:161]: You can also use the exposed event count methods directly
```suggestion
        self.assertEqual(rawio.event_count(0, 0, 0), 1)
        self.assertEqual(rawio.event_count(0, 1, 0), 2)
        self.assertEqual(rawio.event_count(0, 2, 0), 2)
```

**Review by JuliaSprenger — COMMENTED** [review]
Hi @dancrepeau!
Thanks for reworking the IO a second time. Now the structure and mapping to neo streams looks perfect.
I left a couple of small suggestions, comments and questions.
It also seems that you are only considering a part of `record` data to be loaded into events if these are of type `Note` or `Nlx` - is there potentially more data in those records that could be exposed via Neo?
Currently the tests are still failing due to a segmentation fault. Do you get the same error on your system and do you have any idea what's going wrong there?

**Comment by dancrepeau** [comment]
Hi @JuliaSprenger!

As always, thanks for the feedback!  I will look into the segfault in the next day or two and try to figure out what is going on.  I do most of my testing on Mac (M1 chip) and everything is passing for me.  Since this appears to be segfaulting on Ubuntu I will try to reproduce the problem there.

**Comment by JuliaSprenger** [comment]
Hi @dancrepeau Any news on the segfault? Do you have access to an ubuntu system to test this or is gh actions sufficient for you to tests this?

**Comment by dancrepeau** [comment]
Hi @JuliaSprenger!

Yes, we tracked down the segfault and fixed it.  The newest dhn_med_py version is 1.1.1 which has the fix in it.  Thank you for being patient!

I will go through the rest of your comments and update the MED wrapper in the near future.

thanks!

File: neo/rawio/medrawio.py
Comment (dancrepeau) [reply neo/rawio/medrawio.py:280]: While MED and Neuralynx data are separate formats (and MED is an open-source format), the Dark Horse Neuro Inc. (the authors and maintainers of MED) have partnered in the past with Neuralynx, and one of Dark Horse's core products is an acquisition system that works with Neuralynx hardware.  So a lot of MED users would have an intrinsic need for Neuralynx events.

File: neo/rawio/medrawio.py
Comment (dancrepeau) [reply neo/rawio/medrawio.py:282]: I agree that this may not be the most efficient way to do this, but in MED there are potentially an infinite number of types of events, although only a handful have been defined so far.  In Neo, all events have to be represented as strings, whereas in MED, they can be arbitrary binary structures.   Each event type supported by Neo would need a "to_string" function.
One of the obstacles I ran into with the Neo wrapper, in general, is the distinction between retrieving samples (analogsignal_chunk) and retrieving events.  The MED way of doing this, in the dhn_med_py library, it to retrieve both at the same time with a single call.  This matches the use-case for a data viewer, where you might want 30 seconds of data and the same 30 seconds of events.
For the purposes of the Neo wrapper, we created a separate get_session_records() function within dhn_med_py, which only retrives records for a certain time range.  That's the call being used here.

File: neo/rawio/medrawio.py
Comment (dancrepeau) [reply neo/rawio/medrawio.py:271]: I don't fully understand this - I will need to reserach this a bit.

File: neo/rawio/medrawio.py
Comment (dancrepeau) [reply neo/rawio/medrawio.py:280]: Here is a link to Neuralynx's webpage, with some background information about MED, if anyone is interested:
https://neuralynx.com/home/partnerships

File: neo/rawio/medrawio.py
Comment (dancrepeau) [reply neo/rawio/medrawio.py:282]: For the counting of events in a stream/segment, I could move that loop to the parse header and cache the results (cache the count number).  If you think all the data of the events should be cached, we can do that too - although that is potentially a lot of data, as there could be automated events (spikes, seizure detections, or hardware generated events).  The dhn_med_py library's get_session_records(t1, t2) function is actually pretty fast, if a narrow time slice is specified, since only looks in the relevant sections of data to retrieve the events.  Internally, all MED data is divided into segments (which is a slightly different concept than Neo's segments).  When reading samples or events from MED, only the relevant segments are searched, and in the case of samples, caching is done for already decoded data within the dhn_med_py library to speed it up.

There might be some tricks we could do for counting events (without actually looking at the data in them).  That's not currently a part of dhn_med_py, as we didn't anticipate user need for that.

File: neo/rawio/medrawio.py
Comment (dancrepeau) [reply neo/rawio/medrawio.py:97]: We'll have to look into this one a little more.  There's not currently a mechanism to simply read the total number of samples from a channel during the session open phase.  We could add this feature.  I just mentioned it to my colleague Matt Stead, and we think that with all the current MED acquisition software in use now, there will always be the same number of samples for channels with the same sampling frequency.  It is true that potential future software written by people other than Dark Horse Neuro might not make that assumption.  We'll think about this one a bit.

Comment (dancrepeau) [reply neo/rawio/medrawio.py:97]: Thinking a bit more about this one: we already calculate the number of samples in a stream using a "reference" channel (which is the first channel we find to be in a stream).  We could just do that same calculation of every channel prior to making the determination of whether it should be put into an existing stream.  I do worry it might find cases of a single sample at the end of a long stream not being present, due to manner in which the acquisition was stopped, causing the entire channel to be put in a new stream.  But I don't know of any actual cases where this has happened.

File: neo/rawio/medrawio.py
Comment (dancrepeau) [reply neo/rawio/medrawio.py:249]: I tried what you have here, and I think it's close, but it isn't quite right.  Several tests are failing when I substitute this code for what I had:

AssertionError: 
Arrays are not equal

Mismatched elements: 47244 / 48000 (98.4%)
Max absolute difference: 1954.
Max relative difference: 987.
 x: array([[  -1.,  844.,   -5.],
       [  -4.,   -1.,  868.],
       [  -4.,  869., -849.],...
 y: array([[ -1.,  65., 190.],
       [ -4.,  97.,  69.],
       [ -4.,  35., 127.],...

Comment (dancrepeau) [reply neo/rawio/medrawio.py:249]: Okay, this did work - I just needed to also remove the two lines following, where I was reshaping and transposing the matrix.

**Comment by dancrepeau** [comment]
Hi @JuliaSprenger,

We fixed the segfault in the dhn_med_py library, and I believe I have made the suggested adjustments to the code.  Both RawIO and IO are passing the tests in my local testing.  Any comments are welcome.  Thanks!

File: neo/rawio/medrawio.py
Comment (samuelgarcia) [inline neo/rawio/medrawio.py:246]: I guess theses two lines can b eremoved no ?

File: neo/rawio/medrawio.py
Comment (samuelgarcia) [inline neo/rawio/medrawio.py:248]:
```suggestion
        raw_signals = np.empty((i_stop - i_start, num_channels), dtype='int16')
```

Comment (samuelgarcia) [reply neo/rawio/medrawio.py:248]: or int32

**Comment by samuelgarcia** [comment]
Hi @dancrepeau.
thank you for this.
I made a few comments.
I do not understand totally the neuralynx story of the med format. I let Julia having closer eyes on this part.

File: neo/rawio/medrawio.py
Comment (dancrepeau) [reply neo/rawio/medrawio.py]: Yes, you are correct, they are no longer necessary.  I removed the lines.

File: neo/rawio/medrawio.py
Comment (dancrepeau) [reply neo/rawio/medrawio.py:156]: Yes, that's how I implemented this.  Although MED doesn't support epochs (currently), although it may in the future.

**Comment by dancrepeau** [comment]
Hi @apdavison, as far as I know, it is ready to be merged.  Julia had a question about why MED is using the subset of events that we do (which I think I addressed) and suggested a speed-up for counting the number of events, although the only way I can think to perform that speedup in the parse_header routine would involve duplicating data that is already part of the MedSession python object.

So it's your call - I'm happy with it, and if everyone else is, I'd say let's merge.  Thanks!

File: neo/rawio/medrawio.py
Comment (JuliaSprenger) [reply neo/rawio/medrawio.py:110]: @dancrepeau Does this make sense to you or would you rather keep the code at it is?

File: neo/rawio/medrawio.py
Comment (JuliaSprenger) [reply neo/rawio/medrawio.py:277]: Ok, let's keep the idea of counting the number of events already in the `parse_header` it a need for performance optimisation arises in the future.

File: neo/rawio/medrawio.py
Comment (JuliaSprenger) [reply neo/rawio/medrawio.py:308]: this might be the first bottleneck to target when performance issues arise related to events. Let's keep it in mind for later.

**Review by JuliaSprenger — APPROVED** [review]

**Review by JuliaSprenger — APPROVED** [review]

File: neo/rawio/medrawio.py
Comment (JuliaSprenger) [reply neo/rawio/medrawio.py:110]: ok, let's put this also on the list of potential improvements for the future.

**Pull request merged**
